tests/int/test_headers.py

```diff
@@ -0,0 +1,174 @@
+                'sub': user_id,
+                'auditTrackingId': '91f694e6-3749-42fd-90b0-c3134b0d98f6-1546391',
+                'amr': ['N3_SMARTCARD'],
+                'iss': 'https://am.nhsint.auth-ptl.cis2.spineservices.nhs.uk:443/'
+                       'openam/oauth2/realms/root/realms/NHSIdentity/realms/Healthcare',
+                'tokenName': 'id_token',
+                'aud': '969567331415.apps.national',
+                'c_hash': 'bc7zzGkClC3MEiFQ3YhPKg',
+                'acr': 'AAL3_ANY',
+                'org.forgerock.openidconnect.ops': '-I45NjmMDdMa-aNF2sr9hC7qEGQ',
+                's_hash': 'LPJNul-wow4m6Dsqxbning',
+                'azp': '969567331415.apps.national',
+                'auth_time': 1610559802,
+                'realm': '/NHSIdentity/Healthcare',
+                'exp': int(time()) + 6000,
+                'tokenType': 'JWTToken',
+                'iat': int(time()) - 100
+        }
+
+        print('Claims:')
+        print(claims)
+
+        client_assertion_jwt = oauth.create_jwt(kid='test-1')
+        id_token_jwt = oauth.create_id_token_jwt(algorithm="RS512", kid="identity-service-tests-1", claims=claims, signing_key=contents)
+        # When
+        resp = await oauth.get_token_response(
+            grant_type="token_exchange",
+            _jwt=client_assertion_jwt,
+            id_token_jwt=id_token_jwt
+        )
+        print('JWT Token response:')
+        print(resp)
+
+        return resp['body']['access_token']
+
+
+    @pytest.mark.asyncio
+    async def test_headers(self, product, app):
+        # Assign the new product to the app
+        await app.add_api_product([product.name])
+
+        user_id = '0987654321123'
+        access_code = await self.get_access_code(app, user_id)
+
+        client_request_headers = {
+            HEADER_ECHO : '',
+            HEADER_AUTHORIZATION : 'Bearer ' + access_code,
+            ORIGINAL_HEADERS["REFERRAL_ID"] : '000000040032',
+            ORIGINAL_HEADERS["CORRELATION_ID"] : '123123-123123-123123-123123',
+            ORIGINAL_HEADERS["BUSINESS_FUNCTION"] : 'REFERRING_CLINICIAN',
+            ORIGINAL_HEADERS["ODS_CODE"] : 'R69',
+            ORIGINAL_HEADERS["FILENAME"] : 'mysuperfilename.txt',
+            ORIGINAL_HEADERS["COMM_RULE_ORG"] : 'R100'}
+
+        response = requests.get(BASE_URL, headers=client_request_headers)
+        assert response.status_code == 200, "Expected a 200 when accesing the api but got " + (str)(response.status_code)
+
+        json_response = response.json()
+        target_request_headers = json_response['headers']
+
+        for header in ORIGINAL_HEADERS:
+            assert ORIGINAL_HEADERS[header] not in target_request_headers
+
+        assert HEADER_X_REQUEST_ID not in target_request_headers
+
+        for header in RENAMED_HEADERS:
+            assert RENAMED_HEADERS[header] in target_request_headers
+
+        assert target_request_headers[RENAMED_HEADERS["REFERRAL_ID"]] == '000000040032'
```

Comment:
Can these values be extracted out in to variables so they can be used in both the request and assertions? Saves some duplication and easier to read.

Reply:
They were in their own variables but I moved it out as they are more maintainable that way.

Reply:
How come? As it is now I have to look at both the client_request_headers and the assertions to check they are the same values.

Reply:
Both the same in what way? As original headers and renamed headers are keyed on the same value but indicate the different header names accordingly. 
The main benefit with the original headers and renamed headers is that I can loop through them and verify they are either there or not there. 
The final assertion is just confirming the specific value. It's already been confirmed that a value exists.

Reply:
This could be done with parametrize - https://docs.pytest.org/en/6.2.x/parametrize.html

Reply:
At the most I could extract the keys to constants like "REFERRAL_ID" if that's what would ease your confusion?

Reply:
Oh hang on I mis understood I thought you meant the values like REFERRAL_ID but you mean like 000000040032 ?
I can move them into constants too.

Reply:
I've now pushed a change to move the asserted values into constants as well as some other constants added

Reply:
Sorry no I mean you can use the test functionality pytest offers - parametrize. It means you pass in a variable to the test and extract all these headers to outside the test then pytest automatically runs the test for every expected headers and is great for refactoring. Cleans up the asserts. If there's something that doesn't fit the parametrize it's worth considering if it needs a separate test

Reply:
Yeah that is what I meant @andrewclark2020.

Regarding @acarriedev's comment, I don't think we need this now as currently we only have one set of parameters for assertion. Good to know though.